src/java.base/share/classes/java/math/MutableBigInteger.java

```diff
@@ -1793,93 +1865,166 @@ private boolean unsignedLongCompare(long one, long two) {
-     * overflows the range of {@code int}.
-     * @return the integer square root of {@code this}
-     * @since 9
+     * @return the integer square root of {@code this} and the remainder if needed
      */
-    MutableBigInteger sqrt() {
+    MutableBigInteger[] sqrtRem(boolean needRemainder) {
         // Special cases.
-        if (this.isZero()) {
-            return new MutableBigInteger(0);
-        } else if (this.value.length == 1
-                && (this.value[0] & LONG_MASK) < 4) { // result is unity
-            return ONE;
-        }
-
-        if (bitLength() <= 63) {
-            // Initial estimate is the square root of the positive long value.
-            long v = new BigInteger(this.value, 1).longValueExact();
-            long xk = (long)Math.floor(Math.sqrt(v));
-
-            // Refine the estimate.
-            do {
-                long xk1 = (xk + v/xk)/2;
-
-                // Terminate when non-decreasing.
-                if (xk1 >= xk) {
-                    return new MutableBigInteger(new int[] {
-                        (int)(xk >>> 32), (int)(xk & LONG_MASK)
-                    });
+        if (this.intLen <= 2) {
+            final long x = this.toLong(); // unsigned
+            long s = ulongSqrt(x);
+
+            return new MutableBigInteger[] {
+                    new MutableBigInteger((int) s),
+                    needRemainder ? new MutableBigInteger(x - s * s) : null
+            };
+        }
+
+        // Normalize
+        MutableBigInteger x = this;
+        int shift = Integer.numberOfLeadingZeros(x.value[x.offset]) & ~1; // shift must be even
+        if ((x.intLen & 1) != 0)
+            shift += 32; // x.intLen must be even
+
+        if (shift != 0) {
+            x = new MutableBigInteger(x);
+            x.leftShift(shift);
+        }
+
+        // Compute sqrt and remainder
+        MutableBigInteger[] sqrtRem = x.sqrtRemZimmermann(x.intLen, needRemainder);
+
+        // Unnormalize
+        if (shift != 0) {
+            final int halfShift = shift >> 1;
+            if (needRemainder) {
+                // shift <= 62, so s0 is at most 31 bit long
+                final long s0 = sqrtRem[0].value[sqrtRem[0].offset + sqrtRem[0].intLen - 1]
+                        & (-1 >>> -halfShift); // Remove excess bits
+                if (s0 != 0L) { // An optimization
+                    MutableBigInteger doubleProd = new MutableBigInteger();
+                    sqrtRem[0].mul((int) (s0 << 1), doubleProd);
+
+                    sqrtRem[1].add(doubleProd);
+                    sqrtRem[1].subtract(new MutableBigInteger(s0 * s0));
                 }
+                sqrtRem[1].rightShift(shift);
+            }
+            sqrtRem[0].primitiveRightShift(halfShift);
+        }
+        return sqrtRem;
+    }
 
-                xk = xk1;
-            } while (true);
-        } else {
-            // Set up the initial estimate of the iteration.
+    private static long ulongSqrt(long x) {
```

Comment:
Since there is a `unisgnedLongCompare` already, it would be better to name this `unsignedLongSqrt`.